Add notifications and signals to the create-app template

mareklibra opened the pull request:
The ` npx @backstage/create-app` will newly include the `notifications` and `signals` plugins by default.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin commented on packages/create-app/templates/default-app/packages/app/src/App.tsx:
```diff
@@ -36,9 +36,12 @@ import { AppRouter, FlatRoutes } from '@backstage/core-app-api';
 
 const app = createApp({
   apis,
+  plugins: Object.values(plugins),
```
Why do we need this? Generally we have moved away from needing to do this.

mareklibra replied:
@awanlin Can you please suggest proper replacement?

christophe-f replied:
Any updates on this?

awanlin replied:
Not really, the 3 items I highlighted in my review comment have not been addressed yet which I see as blockers from this moving forward. 

For this specific comment I don't know an alternative, in most plugins it just works without needing this. Not sure why that isn't the case for the Notifications plugin.

@Rugvip any thoughts on this?

mareklibra replied:
The signals FE plugin is not discovered otherwise. It has been long time since I was looking into it. IIRC, the signals plugin has no component to be attached into the tree. But I might be wrong now.

mareklibra replied:
> Not really, the 3 items I highlighted in my review comment have not been addressed yet which I see as blockers from this moving forward.

This effort is not abandoned, just postponed due to recent higher-priority tasks. I will come back to it later, will start with the documentation.

awanlin replied:
This has been resolved with the addition of `<SignalsDisplay />`

awanlin reviewed (changes requested):
Hi @mareklibra, this over all looks good to me, left one comment.

I do have more questions and some concerns:

1. The BEP for this is currently still set as provisional, should we be adding features to the default while it is in that state? This might just be a simple update to that BEP.
2. There is no documentation at all on Notifications or Signals. If we add this to the default there will be questions about it with nothing to point people towards or for them to learn about on their own.
3. Related to the above, what notifications will be triggered or available out-of-the-box with this as part of the default? If there are none does it make sense to have this ship as part of the default at this time?

Sorry, I do genuinely want to see this move forward but would like to avoid the issues we had when we changed the default to the new backend system. Notifications are a long time request of the Community and I know that once it becomes default there will be lots of questions. We just need to be prepared for that.

mareklibra commented:
Rebased and updated for documentation.

@awanlin @Rugvip @freben Can you please have a look? This is first me adding a section to the Backstage documentation. Would you rephrase it? Emphasize different aspects?

mareklibra commented:
> The BEP for this is currently still set as provisional, should we be adding features to the default while it is in that state? This might just be a simple update to that BEP.

https://github.com/backstage/backstage/pull/25407

mareklibra commented:
> Related to the above, what notifications will be triggered or available out-of-the-box with this as part of the default? If there are none does it make sense to have this ship as part of the default at this time?

Considering recent state of this PR, the notifications "system" is fully functional from the very beginning. Plugins  or external services can start sending them right after installation. From the very beginning, the notifications were meant to become the core part of every Backstage deployment.

We might think about emitting notifications from the flows available right after `create-app`, but they are very limited at this moment.

In my opinion, it's ok to leave it as it is. 

Or if we don't come up a good use-case for the other core-plugins to emit something, what about sending a broadcast message after "fresh" installation? It can be either a "Welcome" user message or a list of a few technical details about the deployment.

mareklibra commented:
@Rugvip Can you please recommend better way how to discover the `signals` plugin, please?
https://github.com/backstage/backstage/pull/24963/files#diff-c6453d0dd45f236e992074aae7ead2217f5fc579a4a3420f2ed66af9a02c7df2R44

awanlin commented:
Thanks for the updates @mareklibra, I'm going to give this a review on Friday as that's the soonest I'll have time. I do have an early suggestions now that I have seen your docs. Would it make more sense to break these docs out into their own PR that way we can merge them in sooner? They seem foundational and if you are saying it should all work as of the `1.28.x` release of Backstage then that makes more sense to me.

mareklibra commented:
Documentation split: https://github.com/backstage/backstage/pull/25413

awanlin commented:
> @Rugvip Can you please recommend better way how to discover the `signals` plugin, please? https://github.com/backstage/backstage/pull/24963/files#diff-c6453d0dd45f236e992074aae7ead2217f5fc579a4a3420f2ed66af9a02c7df2R44

Sorry @mareklibra `@Rugvip` is on vacation, I'll see if I can follow up with the team on this item but this might be a problem generally, lots of vacation coming up.

mareklibra commented:
Not stale, blocked on https://github.com/backstage/backstage/pull/25413. And needs rebase.

mareklibra commented:
Rebased

mareklibra commented:
Rebased

pkliczewski commented:
@awanlin @Rugvip is there anything else you want Marek to fix in this PR?

Rugvip reviewed (changes requested):
Alright, couple comments for moving this forward.

Tbh my main concern with this is that there is no built-in usage of notifications out of the box, meaning it's a bit of a dead feature in new apps unless adopters pick it up and send their own notifications. It does promote the feature though so it's not without benefit, and it's very much a system-wide feature that makes sense to eventually include in the app.

Really where I'd be at at this point is that I think it's best if we can close out the BEP before adding this to the app, so that we know that we're at a reasonably stable point with the APIs.

mareklibra commented on packages/create-app/templates/default-app/packages/app/src/components/Root/Root.tsx:
```diff
@@ -75,11 +76,13 @@ export const Root = ({ children }: PropsWithChildren<{}>) => (
         <SidebarScrollWrapper>
           {/* Items in this group will be scrollable if they run out of space */}
         </SidebarScrollWrapper>
+        <SidebarDivider />
```
removed

mareklibra commented:
Rebased on top of the #26200. Waiting for `plugins-signal` release

awanlin commented on packages/create-app/templates/default-app/packages/app/src/components/Root/Root.tsx:
```diff
@@ -75,6 +76,7 @@ export const Root = ({ children }: PropsWithChildren<{}>) => (
         <SidebarItem icon={ExtensionIcon} to="api-docs" text="APIs" />
         <SidebarItem icon={LibraryBooks} to="docs" text="Docs" />
         <SidebarItem icon={CreateComponentIcon} to="create" text="Create..." />
+        <NotificationsSidebarItem />
```
I'm wondering about the position of this: does it make sense to be here or should we do like @drodil did for the Demo site? https://demo.backstage.io/

Code example:

https://github.com/backstage/demo/blob/846e2be70369053b50e2bdec42df68449e2aff32/packages/app/src/components/Root/Root.tsx#L104-L106

mareklibra replied:
Np, done:

![Screenshot from 2024-09-24 09-26-00](https://github.com/user-attachments/assets/2fddb682-39ff-4008-a507-d62c91a50201)

awanlin reviewed (changes requested):
Getting closer, just left one comment about the location in the sidebar, just want to be sure we get this right from the start 👍

awanlin reviewed (approved):
Awesome! 🚀

awanlin commented:
Hi @Rugvip, can you give this another look? Pretty sure it's ready to go! Would be nice to see it in one of the `next` releases before the upcoming `main` release.

mareklibra commented:
Not stale, pending review

awanlin commented:
Sorry for the ping @Rugvip, just need your approval to move this one forward 👍

mareklibra commented:
Still interested in merging that

mareklibra commented:
All I want for Christmas is...

awanlin commented:
HI @mareklibra, I'd very much like to push to get this into the `1.37.0` April release. Can you rebase your branch, please?

Rugvip reviewed (commented):
Just to keep this ready to merge: we'll need to install the events plugin backend as well, so that scaled deployments with the signals plugin works correctly.

mareklibra commented:
Added events-backend.

@Rugvip Can you please have a look?
Wdyt about adding an empty section for events to the app-config?

awanlin commented:
Hi @mareklibra, I think if you rebase it should get the CI green once again 👍
